src/components/icon/icon-provider.ts

```diff
@@ -0,0 +1,255 @@
+import {Injectable, Renderer} from 'angular2/core';
+import {Http, Response, HTTP_PROVIDERS} from 'angular2/http';
+import {AsyncSubject, Observer, Observable} from 'rxjs/Rx';
+
+/**
+  *  Configuration for an icon, possibly including the cached SVG element.
+  */
+class IconConfig {
+  svgElement: SVGElement = null;
+  constructor(public url: string, public viewBoxSize: number) {
+  }
+}
+
+@Injectable()
+export class MdIconProvider {
+  // IconConfig objects and cached SVG elements for individual icons.
+  // First level of cache is icon set (which is the empty string for the default set).
+  // Second level is the icon name within the set.
+  private _iconConfigs = new Map<string, Map<string, IconConfig>>();
+
+  // IconConfig objects and cached SVG elements for icon sets.
+  // These are stored only by set name, but multiple URLs can be registered under the same name.
+  private _iconSetConfigs = new Map<string, [IconConfig]>();
+
+  // Cache for icons loaded by direct URLs.
+  private _cachedIconsByUrl = new Map<string, SVGElement>();
+
+  private _fontClassNamesByAlias = new Map<string, string>();
+
+  private _inProgressUrlFetches = new Map<string, Observable<string>>();
+  
+  private _defaultViewBoxSize = 24;
+  private _defaultFontSetClass = 'material-icons';
+  
+  constructor(private _http: Http) {
+  }
+  
+  public addIcon(iconName: string, url: string, viewBoxSize:number=0): MdIconProvider {
+    return this.addIconInSet('', iconName, url, viewBoxSize);
+  }
+  
+  public addIconInSet(
```

Comment:
I think we can separate the concept of icon namespaces and icon sets. We can treat an icon-set is a bundle of icons all in one file, which can be added to any namespace. Perhaps it makes sense to just use the term "icon bundle".

```ts
addIcon(name, url, opt_namespace)

addIconBundle(url, opt_namespace)
```

Then again, maybe it does make more sense to _always_ treat a "bundle" as a namespace. Let me run this idea by the rest of the team.